### PR TITLE
Add x86 runtime for testing

### PR DESCRIPTION
**Pull request by hoyosjs**
Contributes to dotnet/arcade#5583

## Proposed changes

- Install x86 runtime for testing.

## Customer Impact

- Ensures proper testing in a 32 bit process.

## Regression? 

- No

## Risk

- None, other than increased download at first build and potential test failures from the increased coverage.

### After

image

## Test methodology 

- This adds no functionality so no added tests.

## Test environment(s) 

- Using default runtimes and SDK already used for testing.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3396)

### CODE REVIEW

File: global.json [hunk 1]
```diff
@@ -2,9 +2,8 @@
   "tools": {
     "dotnet": "5.0.100-preview.6.20266.3",
     "runtimes": {
-      "dotnet": [
-        "$(MicrosoftNETCoreAppPackageVersion)"
-      ]
+      "dotnet/x64": [ "$(MicrosoftNETCoreAppPackageVersion)" ],
```
Comment (weltkante) [inline global.json:5]: @RussKie does this change also require updating [`start-vs.cmd`](https://github.com/dotnet/winforms/blob/master/start-vs.cmd#L7) and `start-code.cmd` which set environment variables?

Comment (hoyosjs) [reply global.json:5]: Those already set the appropriate variables: https://github.com/dotnet/winforms/blob/1c3c52c6d170db7b98923c264190ea23add129ef/start-vs.cmd#L8

Comment (weltkante) [reply global.json:5]: was asking if the x64 folder also moves into a subdirectory, `"dotnet/x64"` makes it look like that

Comment (RussKie) [reply global.json:5]: > does this change also require updating [`start-vs.cmd`](https://github.com/dotnet/winforms/blob/master/start-vs.cmd?rgh-link-date=2020-06-05T07%3A11%3A28Z#L7) and `start-code.cmd` which set environment variables?

I think it does, if the location of dotnet.exe is changed.

Comment (RussKie) [reply global.json:5]: Would for arm64 we have "dotnet/arm64"?

Comment (RussKie) [reply global.json:5]: I've run it locally, no x64 folder is created, but both x64 and x86 SDKs are installed

```
C:\Development\winforms>restore
Processing C:\Development\winforms\eng\..\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_netcoreapp3.0 was not found in .
Native tools bootstrap failed
Processing C:\Development\winforms\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_netcoreapp3.0 was not found in .
Native tools bootstrap failed
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.6.20266.3/dotnet-sdk-5.0.100-preview.6.20266.3-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-preview.6.20266.3/dotnet-sdk-5.0.100-preview.6.20266.3-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\Development\winforms\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
  Determining projects to restore...
  All projects are up-to-date for restore.
Processing C:\Development\winforms\eng\..\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_netcoreapp3.0 was not found in .
Native tools bootstrap failed
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.7.20303.11/dotnet-runtime-5.0.0-preview.7.20303.11-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.7.20303.11/dotnet-runtime-5.0.0-preview.7.20303.11-win-x64.zip
dotnet-install: Installation finished
Processing C:\Development\winforms\eng\..\global.json
cmake.exe already exists; replacing...
dotnet-api-docs_netcoreapp3.0 was not found in .
Native tools bootstrap failed
GET https://dot.net/v1/dotnet-install.ps1
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.7.20303.11/dotnet-runtime-5.0.0-preview.7.20303.11-win-x86.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-preview.7.20303.11/dotnet-runtime-5.0.0-preview.7.20303.11-win-x86.zip
dotnet-install: Adding to current process PATH: "C:\Development\winforms\.dotnet\x86\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
  Determining projects to restore...
  All projects are up-to-date for restore.
```

image

**Comment by RussKie** [comment]
Looks great 👍 

Now x86 tests are failing, as expected (see #3359):
```
    System.Windows.Forms.Tests.ListViewTests.ListView_Handle_GetTextBackColor_Success [FAIL]
      System.OverflowException : Arithmetic operation resulted in an overflow.
      Stack Trace:
           at System.IntPtr.op_Explicit(Int64 value)
        F:\workspace\_work\1\s\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\ListViewTests.cs(1623,0): at System.Windows.Forms.Tests.ListViewTests.ListView_Handle_GetTextBackColor_Success()
```

**Comment by weltkante** [comment]
I assume you want to fix #3359 before merging this to avoid CI failures, the quickest solution I can come up with is ugly: `Environment.Is64BitProcess ? (IntPtr)0xFFFFFFFF : (IntPtr)(-1)` because 0xFFFFFFFF is no valid 32 bit IntPtr but -1 is sign extended and differs in 64 bit.

File: global.json (hunk 1 above)
Comment (hoyosjs) [reply global.json:5]: > I think it does, if the location of dotnet.exe is changed.

I might have misread them, but it seems like it points to the correct directories. 

> Would for arm64 we have "dotnet/arm64"?

A few things:

- You don't seem to have any ARM64 hosts to run tests in ATM in CI or anything. However the ARM64 runtime would always get downloaded to the build host. It makes sense for x64/x86 as it actually tests both. However, I don't know if you want every user to download an ARM64 runtime. The keys will only really tell it to download x86 or x64 (https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/src/InstallDotNetCore.cs)
- You'd need to deploy the build and a runtime to the test machine. For the `build.cmd -test` thing to work out you'd also need an SDK. The runtime key of the global.json only really fetches runtimes, so even it you got an install, you'd need.
- The MSBuild logic that sets up the parameters for testing only accounts for x86 and x64. The few repositories that use this don't use it to test ARM64 as far as I know (AspNet, CoreCLR and CoreFX in runtime), but rather solutions like corerun based runs of xunit or calls to vstest.

**Comment by RussKie** [comment]
I'm working on a fix, should have it shortly

File: global.json (hunk 1 above)
Comment (hoyosjs) [reply global.json:5]: > I've run it locally, no x64 folder is created, but both x64 and x86 SDKs are installed

Yeah, that's expected. `.dotnet` actually contains the x64 SDK and runtime. x86 contains a host and a runtime (no SDK).

**Comment by RussKie** [comment]
Thank you @hoyosjs

**Pull request merged**